src/GafferImage/Mix.cpp

```diff
@@ -0,0 +1,291 @@
+		}
+	}
+}
+
+IECore::ConstStringVectorDataPtr Mix::computeChannelNames( const Gaffer::Context *context, const ImagePlug *parent ) const
+{
+	IECore::StringVectorDataPtr outChannelStrVectorData( new IECore::StringVectorData() );
+	std::vector<std::string> &outChannels( outChannelStrVectorData->writable() );
+
+	for( ImagePlugIterator it( inPlugs() ); !it.done(); ++it )
+	{
+		if( (*it)->getInput<ValuePlug>() )
+		{
+			IECore::ConstStringVectorDataPtr inChannelStrVectorData((*it)->channelNamesPlug()->getValue() );
+			const std::vector<std::string> &inChannels( inChannelStrVectorData->readable() );
+			for ( std::vector<std::string>::const_iterator cIt( inChannels.begin() ); cIt != inChannels.end(); ++cIt )
+			{
+				if ( std::find( outChannels.begin(), outChannels.end(), *cIt ) == outChannels.end() )
+				{
+					outChannels.push_back( *cIt );
+				}
+			}
+		}
+	}
+
+	if ( !outChannels.empty() )
+	{
+		return outChannelStrVectorData;
+	}
+
+	return inPlug()->channelNamesPlug()->defaultValue();
+}
+
+void Mix::hashChannelData( const GafferImage::ImagePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ImageProcessor::hashChannelData( output, context, h );
+
+	const std::string channelName = context->get<std::string>( ImagePlug::channelNameContextName );
+	const V2i tileOrigin = context->get<V2i>( ImagePlug::tileOriginContextName );
+	const Box2i tileBound( tileOrigin, tileOrigin + V2i( ImagePlug::tileSize() ) );
+
+
+	for( ImagePlugIterator it( inPlugs() ); !it.done(); ++it )
+	{
+		if( !(*it)->getInput<ValuePlug>() )
+		{
+			continue;
+		}
+
+		IECore::ConstStringVectorDataPtr channelNamesData = (*it)->channelNamesPlug()->getValue();
+		const std::vector<std::string> &channelNames = channelNamesData->readable();
+
+		if( ImageAlgo::channelExists( channelNames, channelName ) )
+		{
+			(*it)->channelDataPlug()->hash( h );
+		}
+
+
+		// The hash of the channel data we include above represents just the data in
+		// the tile itself, and takes no account of the possibility that parts of the
+		// tile may be outside of the data window. This simplifies the implementation of
+		// nodes like Constant (where all tiles are identical, even the edge tiles) and
+		// Crop (which does no processing of tiles at all). For most nodes this doesn't
+		// matter, because they don't change the data window, or they use a Sampler to
+		// deal with invalid pixels. But because our data window is the union of all
+		// input data windows, we may be using/revealing the invalid parts of a tile. We
+		// deal with this in computeChannelData() by treating the invalid parts as black,
+		// and must therefore hash in the valid bound here to take that into account.
+		const Box2i validBound = boxIntersection( tileBound, (*it)->dataWindowPlug()->getValue() );
+		h.append( validBound );
+	}
+
+	const std::string &maskChannel = maskChannelPlug()->getValue();
+	h.append( maskPlug()->channelDataHash( maskChannel, tileOrigin ) );
```

Comment:
We need to check `ImageAlgo::channelExists( maskPlug(), maskChannel )` here before calling `channelDataHash()`.